presto-main/src/test/java/com/facebook/presto/operator/TestExchangeClient.java

```diff
@@ -320,6 +321,80 @@ public void testClose()
+    @Test
+    public void testRemoveRemoteSource()
+            throws Exception
+    {
+        DataSize maxResponseSize = new DataSize(1, Unit.BYTE);
+        MockExchangeRequestProcessor processor = new MockExchangeRequestProcessor(maxResponseSize);
+
+        URI location1 = URI.create("http://localhost:8081/foo");
+        TaskId taskId1 = TaskId.valueOf("foo");
+        URI location2 = URI.create("http://localhost:8082/bar");
+        TaskId taskId2 = TaskId.valueOf("bar");
+
+        processor.addPage(location1, createPage(1));
+        processor.addPage(location1, createPage(2));
+        processor.addPage(location1, createPage(3));
+
+        ExchangeClient exchangeClient = new ExchangeClient(
+                new DataSize(1, Unit.BYTE),
+                maxResponseSize,
+                1,
+                new Duration(1, TimeUnit.MINUTES),
+                true,
+                new TestingHttpClient(processor, newCachedThreadPool(daemonThreadsNamed("test-%s"))),
+                scheduler,
+                new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
+                pageBufferClientCallbackExecutor);
+        exchangeClient.addLocation(location1, taskId1);
+        exchangeClient.addLocation(location2, taskId2);
+
+        // fetch a page
+        assertEquals(exchangeClient.isClosed(), false);
+        assertPageEquals(getNextPage(exchangeClient), createPage(1));
+
+        // remove remote source while pages are still available
+        exchangeClient.removeRemoteSource(taskId1);
+
+        // client should not receive any further pages from removed remote source
+        assertNull(exchangeClient.pollPage());
+        assertEquals(exchangeClient.getStatus().getBufferedPages(), 0);
+
+        // add pages to another source
+        processor.addPage(location2, createPage(4));
+        processor.addPage(location2, createPage(5));
+        processor.addPage(location2, createPage(6));
+        processor.setComplete(location2);
+
+        assertEquals(exchangeClient.isClosed(), false);
+        assertPageEquals(getNextPage(exchangeClient), createPage(4));
+        assertEquals(exchangeClient.isClosed(), false);
+        assertPageEquals(getNextPage(exchangeClient), createPage(5));
+        assertEquals(exchangeClient.isClosed(), false);
+        assertPageEquals(getNextPage(exchangeClient), createPage(6));
+
+        assertFalse(tryGetFutureValue(exchangeClient.isBlocked(), 10, MILLISECONDS).isPresent());
+        assertEquals(exchangeClient.isClosed(), false);
+
+        exchangeClient.noMoreLocations();
+        // The transition to closed may happen asynchronously, since it requires that all the HTTP clients
+        // receive a final GONE response, so just spin until it's closed or the test times out.
+        while (!exchangeClient.isClosed()) {
+            Thread.sleep(1);
+        }
+
+        PageBufferClientStatus clientStatus1 = exchangeClient.getStatus().getPageBufferClientStatuses().get(0);
```

Comment:
You can also extract a private method `assertPageBufferClientStatus` instead of copying this code, and call it twice here.